opm/simulators/linalg/bda/BILU0.cpp

```diff
@@ -129,6 +153,233 @@ namespace bda
+                int iRowEnd = Lmat->rowPointers[row + 1];
+
+                for (int ij = iRowStart; ij < iRowEnd; ij++) {
+                    int j = Lmat->colIndices[ij];
+                    // refine Lij element
+                    int i1 = LUmat->rowPointers[row];
+                    int i2 = LUmat->rowPointers[row+1];
+                    int kk = 0;
+                    for(kk = i1; kk < i2; ++kk) {
+                        ptrdiff_t c = LUmat->colIndices[kk];
+                        if (c >= j) {
+                            break;
+                        }
+                    }
+                    double aij[bs*bs];
+                    memcpy(&aij[0], LUmat->nnzValues + kk * bs * bs, sizeof(double) * bs * bs);
+                    int jk = Ut->rowPointers[j];
+                    int ik = Ut->colIndices[jk];
+                    for (int k = iRowStart; k < ij; ++k) {
+                        int ki = Lmat->colIndices[k];
+                        while(ik < ki) {
+                            ++jk;
+                            ik = Ut->colIndices[jk];
+                        }
+
+                        if(ik == ki) {
+                            blockMultSub<bs>(&aij[0], Lmat->nnzValues + k * bs * bs , Ut->nnzValues + jk * bs * bs);
+                        }
+                    }
+                    // calculate aij / ujj
+                    double ujj[bs*bs];
+                    inverter(Ut->nnzValues + (Ut->rowPointers[j+1] - 1) * bs * bs, &ujj[0]);
+                    // lij = aij / ujj
+                    blockMult<bs>(&aij[0], &ujj[0], Ltmp + ij * bs * bs);
+
+                }
+            }
+            double *t = Lmat->nnzValues;
+            Lmat->nnzValues = Ltmp;
+            Ltmp = t;
+            t = Ut->nnzValues;
+            Ut->nnzValues = Utmp;
+            Utmp = t;
+        } // end sweep
+        delete[] Ltmp;
+#endif
+
+        // convert Ut to BSR
+        // diagonal stored separately
+        std::vector<int> ptr(Nb+1, 0);
+        std::vector<int> col(Ut->rowPointers[Nb]);
+
+        // count blocks per row for U (BSR)
+        // store diagonal in invDiagVals
+        for(int i = 0; i < Nb; ++i) {
+            for(int k = Ut->rowPointers[i]; k < Ut->rowPointers[i+1]; ++k) {
+                int j = Ut->colIndices[k];
+                if (j == i) {
+                    inverter(Ut->nnzValues + k * bs * bs, invDiagVals + i * bs * bs);
+                } else {
+                    ++ptr[j+1];
+                }
+            }
+        }
+
+        // prefix sum
+        int sumU = 0;
+        for (int i = 1; i < Nb+1; i++) {
+            sumU += ptr[i];
+            ptr[i] = sumU;
+        }
+
+        // actually copy nonzero values for U
+        for(int i = 0; i < Nb; ++i) {
+            for(int k = Ut->rowPointers[i]; k < Ut->rowPointers[i+1]; ++k) {
+                int j = Ut->colIndices[k];
+                if (j != i) {
+                    int head = ptr[j]++;
```

Comment:
??? Like last time, why is this increased twice for off diagonals?

Reply:
It is (mis)used like `Ut->rowPointers` to indicate the position of the next block (global block counter). The first loop calculates the number of blocks per row. After the partial_sum, the values in `ptr` indicate the index of the first block of the row.
One possible way to optimize is to store the rowPointers of U beforehand. So that the first loop and the partial_sum can be removed. I tried this but only got wrong values.

Reply:
Nevermind